From: Frédéric Lécaille <flecaille@haproxy.com>
Date: Fri, 12 May 2017 07:57:15 +0000 (+0200)
Subject: BUG/MAJOR: dns: Broken kqueue events handling (BSD systems).
X-Git-Url: http://git.haproxy.org/?p=haproxy.git;a=commitdiff_plain;h=64920538fc19f3c7f94dfa1e84a9a6569b8c3d37

BUG/MAJOR: dns: Broken kqueue events handling (BSD systems).

Some DNS related network sockets were closed without unregistering their file
descriptors from their underlying kqueue event sets. This patch replaces calls to
close() by fd_delete() calls to that to delete such events attached to DNS
network sockets from the kqueue before closing the sockets.

The bug was introduced by commit 26c6eb8 ("BUG/MAJOR: dns: restart sockets
after fork()") which was backported in 1.7 so this fix has to be backported
there as well.

Thanks to Jim Pingle who reported it and indicated the faulty commit, and
to Lukas Tribus for the trace showing the bad file descriptor.
---

diff --git a/src/dns.c b/src/dns.c
index a118598..cb0a9a9 100644
--- src/dns.c
+++ src/dns.c
@@ -1004,7 +1004,7 @@ int dns_init_resolvers(int close_socket)
 
 			if (close_socket == 1) {
 				if (curnameserver->dgram) {
-					close(curnameserver->dgram->t.sock.fd);
+					fd_delete(curnameserver->dgram->t.sock.fd);
 					memset(curnameserver->dgram, '\0', sizeof(*dgram));
 					dgram = curnameserver->dgram;
 				}
